opal/mca/common/ucx : assert fix -  change thread mode sent to UCX api

nathanb2111 opened the pull request:
UCP Worker Thread Modes in OpenUCX define how multiple threads interact with a UCP worker. `UCS_THREAD_MODE_SINGLE` indicates that only one unique thread can use this worker throughout the whole application. 
In a multithreaded application, OMPI may reuse `winfo` objects among different threads, conflicting with hard-coded `UCS_THREAD_MODE_SINGLE`, and resulting in an assert failure

If an application use multiple threads, we want to use parameter `UCS_THREAD_MODE_SERIALIZED` that allows different threads to use the same worker successively

nbellalou committed:
opal/mca/common/ucx : assert fix -  change thread mode sent to UCX api

Signed-off-by: Nathan Bellalou <nbellalou@nvidia.com>

nbellalou commented:
@gleon99

edgargabriel commented on opal/mca/common/ucx/common_ucx_wpool.c:
```diff
@@ -55,7 +57,7 @@ static opal_common_ucx_winfo_t *_winfo_create(opal_common_ucx_wpool_t *wpool)
         memset(&worker_params, 0, sizeof(worker_params));
         worker_params.field_mask = UCP_WORKER_PARAM_FIELD_THREAD_MODE;
-        worker_params.thread_mode = UCS_THREAD_MODE_SINGLE;
+        worker_params.thread_mode = ompi_mpi_thread_provided == MPI_THREAD_SINGLE ? UCS_THREAD_MODE_SINGLE : UCS_THREAD_MODE_SERIALIZED;
```
if the user initialized MPI with MPI_THREAD_MULTIPLE, will this be passed correctly to the initialization of the UCX worker when creating a window? I am not sure this fix here correctly captures this (or fixes it), see https://github.com/open-mpi/ompi/issues/13347

nathanb2111 commented:
bot:retest

edgargabriel reviewed (dismissed):
Sounds good, I tested the patch against the software that triggered the issue https://github.com/open-mpi/ompi/issues/13347, and I can confirm that the problem seems to be resolved with this patch.

edgargabriel commented:
@janjust once this PR is merged to main, we should backport this to 5.0.x and have it as part of 5.0.10 if possible (and 6.0.x of course)

hppritcha commented:
stuff in opal shouldn't be referencing  ompi_ routines or globals so this does look very legit.

jsquyres reviewed (dismissed):
Need to fix the abstraction violation.

nbellalou committed:
Fix abstraction violation between ompi and opal

Create two bool variables, opal_single_threaded and
opal_common_ucx_single_threaded, that mimic behavior of variables
opal_uses_threads and opal_common_ucx_single_threaded, in order to
propagate mpi thread level to opal while preserving abstraction.
opal_single_threaded is true if and only if mpi thread level is MPI_THREAD_SINGLE

Signed-off-by: Nathan Bellalou <nbellalou@nvidia.com>

jsquyres commented:
@nathanb2111 @nbellalou @janjust If you want this in the v5.0.x or v6.0.x branches, don't forget to open a backport PR.

edgargabriel commented:
> @nathanb2111 @nbellalou @janjust If you want this in the v5.0.x or v6.0.x branches, don't forget to open a backport PR.

I think it was backported to 5.0.x, but not to the 6.0.x branch, we need to do that.

jsquyres commented:
> I think it was backported to 5.0.x, but not to the 6.0.x branch, we need to do that.

@edgargabriel If you want it in v6.0.x, it's a simple backport bot away.  Backporting this PR will also simplify https://github.com/open-mpi/ompi/pull/14209.  That PR is currently not a direct cherry-pick because this PR is not on `v6.0.x`.

nbellalou commented:
@jsquyres @edgargabriel what do you need me to do ?

jsquyres commented:
If you want this put on the v6.0.x branch, then you can either use the backport bot or just make a v6.0.x PR with the cherry-picks of the 2 commits on this PR back to the `v6.0.x` branch.  Do me a favor and let me know if you do this, because that affects my v6.0.x PR #14209

nbellalou commented:
@jsquyres I opened a PR to v6.x.0: https://github.com/open-mpi/ompi/pull/14213

jsquyres commented:
> @jsquyres I opened a PR to v6.x.0: #14213

Perfect; thanks.  We'll get that merged and then I'll rebase #14209 on top of it.  Thanks!
